icu4c/source/i18n/formatted_string_builder.cpp

```diff
@@ -257,16 +258,37 @@ int32_t FormattedStringBuilder::prepareForInsert(int32_t index, int32_t count, U
     U_ASSERT(count >= 0);
     if (index == 0 && fZero - count >= 0) {
         // Append to start
+        int32_t newLength;
+        // Add and return U_INPUT_TOO_LONG_ERROR if overflow int32_t.
+        if (uprv_add32_overflow(fLength, count, &newLength)) {
+          status = U_INPUT_TOO_LONG_ERROR;
+          return count;
+        }
         fZero -= count;
-        fLength += count;
+        fLength = newLength;
         return fZero;
```

Comment:
You should not need to change this branch at all, because the condition `fZero - count >= 0` means that count is small (count <= fZero), and the new contents fits into the existing early part of the buffer. Please revert this part.

Reply:
we only know count is at most fZero per your logic, right? How could we be sure fLength + fZero won't overflow int32_t ?

Reply:
done